executor/builder.go

```diff
@@ -531,27 +531,37 @@ func (b *executorBuilder) buildMergeJoin(v *plan.PhysicalMergeJoin) Executor {
 		joinKeys: rightKeys,
 	}
 
-	mergeJoinExec := &MergeJoinExec{
+	lhsColTypes := leftExec.Schema().GetTypes()
+	rhsColTypes := rightExec.Schema().GetTypes()
+	allColTypes := append(lhsColTypes, rhsColTypes...)
+
+	e := &MergeJoinExec{
 		baseExecutor:    newBaseExecutor(v.Schema(), b.ctx, leftExec, rightExec),
-		resultGenerator: newJoinResultGenerator(b.ctx, v.JoinType, false, v.DefaultValues, v.OtherConditions),
+		resultGenerator: newJoinResultGenerator(b.ctx, v.JoinType, false, v.DefaultValues, v.OtherConditions, allColTypes),
 		stmtCtx:         b.ctx.GetSessionVars().StmtCtx,
 		// left is the outer side by default.
 		outerKeys: leftKeys,
 		innerKeys: rightKeys,
 		outerIter: leftRowBlock,
 		innerIter: rightRowBlock,
 	}
+
+	innerColTypes := rhsColTypes
 	if v.JoinType == plan.RightOuterJoin {
-		mergeJoinExec.outerKeys, mergeJoinExec.innerKeys = mergeJoinExec.innerKeys, mergeJoinExec.outerKeys
-		mergeJoinExec.outerIter, mergeJoinExec.innerIter = mergeJoinExec.innerIter, mergeJoinExec.outerIter
+		e.outerKeys, e.innerKeys = e.innerKeys, e.outerKeys
+		e.outerIter, e.innerIter = e.innerIter, e.outerIter
+		innerColTypes = lhsColTypes
 	}
 
 	if v.JoinType != plan.InnerJoin {
-		mergeJoinExec.outerFilter = mergeJoinExec.outerIter.filter
-		mergeJoinExec.outerIter.filter = nil
+		e.outerFilter = e.outerIter.filter
+		e.outerIter.filter = nil
+		if v.JoinType == plan.LeftOuterJoin || v.JoinType == plan.RightOuterJoin {
+			e.resultGenerator.initDefaultChunkInner(innerColTypes)
```

Comment:
We already have `v.JoinType` in `newJoinResultGenerator`, we can init default chunk inner internally.